File: native/ui/macosx/progressbar.m
#include <Cocoa/Cocoa.h>
#include <assert.h>

#include <chisel-native.h>
#include <chisel-native-ui.h>

#include <chisel-native-view.h>
#include <chisel-native-progressbar.h>

@interface ChiselProgressIndicator : NSProgressIndicator
{
	bool _isAnimating;
}
- (id)initWithFrame:(NSRect)frameRect;
- (void)startAnimation:(id)sender;
- (void)stopAnimation:(id)sender;
- (bool)isAnimating;
@end

@implementation ChiselProgressIndicator
- (id)initWithFrame:(NSRect)frameRect {
	id res = [super initWithFrame:frameRect];
	
	_isAnimating = NO;
	
	return res;
}

- (void)startAnimation:(id)sender {
	if ( !_isAnimating ) {
		_isAnimating = YES;
		[super startAnimation: sender];
	}
}

- (void)stopAnimation:(id)sender {
	if ( _isAnimating ) {
		_isAnimating = NO;
		[super stopAnimation: sender];
	}
}

- (bool)isAnimating {
	return _isAnimating;
}
@end

native_handle _chisel_native_progressbar_create( int direction ) {
	NSProgressIndicator *view = [[ChiselProgressIndicator alloc] init];
	assert( view != nil );
	
	return (native_handle)view;
}

void _chisel_native_progressbar_set_minimum( native_handle native, CLFloat minValue ) {
	NSProgressIndicator *progressbar = (NSProgressIndicator *)native;
	
	[progressbar setMinValue:minValue];
	
	[progressbar displayIfNeeded];
}

void _chisel_native_progressbar_set_maximum( native_handle native, CLFloat maxValue ) {
	NSProgressIndicator *progressbar = (NSProgressIndicator *)native;
	
	[progressbar setMaxValue:maxValue];
	
	[progressbar displayIfNeeded];
}

void _chisel_native_progressbar_set_value( native_handle native, CLFloat value ) {
	NSProgressIndicator *progressbar = (NSProgressIndicator *)native;
	
	double oldVal = [progressbar doubleValue];
	if ( oldVal == [progressbar minValue] || oldVal == [progressbar maxValue] ) {
		[progressbar setIndeterminate: YES];
		[progressbar setIndeterminate: NO];
	}
	
	[progressbar setDoubleValue: value];
	
	[progressbar displayIfNeeded];
}

CLFloat _chisel_native_progressbar_get_value( native_handle native ) {
	NSProgressIndicator *progressbar = (NSProgressIndicator *)native;
	
	return [progressbar doubleValue];
}

CLFloat _chisel_native_progressbar_get_thickness( native_handle native ) {
	return (CLFloat)NSProgressIndicatorPreferredThickness;
}

void _chisel_native_progressbar_set_indeterminate( native_handle native, int val ) {
	NSProgressIndicator *progressbar = (NSProgressIndicator *)native;
	
	[progressbar setIndeterminate: val == 0 ? NO : YES];
	
	[progressbar displayIfNeeded];
}

int _chisel_native_progressbar_get_indeterminate( native_handle native ) {
	NSProgressIndicator *progressbar = (NSProgressIndicator *)native;
	
	return [progressbar isIndeterminate] ? 1 : 0;
}

void _chisel_native_progressbar_set_animating( native_handle native, int val ) {
	NSProgressIndicator *progressbar = (NSProgressIndicator *)native;
	
	if ( val ) {
		[progressbar startAnimation: progressbar];
	} else {
		[progressbar stopAnimation: progressbar];
	}
	
	[progressbar displayIfNeeded];
}

int _chisel_native_progressbar_get_animating( native_handle native ) {
	ChiselProgressIndicator *progressbar = (ChiselProgressIndicator *)native;
	
	return [progressbar isAnimating] ? 1 : 0;
}

File: native/ui/gtk/progressbar.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

#include <gtk/gtk.h>

#include <chisel-native.h>
#include <chisel-native-ui.h>

#include <chisel-native-view.h>
#include <chisel-native-progressbar.h>

#include <chisel-native-enums.h>

#include "widgets.h"

#define ANIMATE_UPDATE_INTERVAL 100

void _chisel_native_progressbar_ind_magic( gpointer data ) {
	GtkWidget *widget = GTK_WIDGET(data);
	
	gtk_progress_bar_pulse( GTK_PROGRESS_BAR(widget) );
	
	return g_object_get_data( G_OBJECT(widget), "chisel-progress-animating" );
}

native_handle _chisel_native_progressbar_create( int direction ) {
	GtkWidget *widget = gtk_progress_bar_new( );
	
	if ( direction == ProgressBarTypeHorizontal ) {
		gtk_progress_bar_set_orientation( GTK_PROGRESS_BAR(widget), GTK_PROGRESS_LEFT_TO_RIGHT );
	} else {
		gtk_progress_bar_set_orientation( GTK_PROGRESS_BAR(widget), GTK_PROGRESS_TOP_TO_BOTTOM );
	}
	
	g_object_set_data( G_OBJECT(widget), "chisel-progress-min", malloc(sizeof(CLFloat)) );
	g_object_set_data( G_OBJECT(widget), "chisel-progress-max", malloc(sizeof(CLFloat)) );
	
	_chisel_native_progressbar_set_minimum( widget, 0 );
	_chisel_native_progressbar_set_minimum( widget, 1 );
	
	_chisel_native_progressbar_set_indeterminate( widget, 1 );
	
	return (native_handle)widget;
}

void _chisel_native_progressbar_set_minimum( native_handle native, CLFloat minValue ) {
	CLFloat *min = g_object_get_data( G_OBJECT(native), "chisel-progress-min" );
	
	*min = minValue;
}

void _chisel_native_progressbar_set_maximum( native_handle native, CLFloat maxValue ) {
	CLFloat *max = g_object_get_data( G_OBJECT(native), "chisel-progress-max" );
	
	*max = maxValue;
}

void _chisel_native_progressbar_set_value( native_handle native, CLFloat value ) {
	CLFloat *min = g_object_get_data( G_OBJECT(native), "chisel-progress-min" );
	CLFloat *max = g_object_get_data( G_OBJECT(native), "chisel-progress-max" );
	
	CLFloat range = *max - *min;
	CLFloat fraction = (value - (*min)) / range;
	
	gtk_progress_bar_set_fraction( GTK_PROGRESS_BAR(native), fraction );
	
	_chisel_force_gtk_refresh( );
}

CLFloat _chisel_native_progressbar_get_value( native_handle native ) {
	CLFloat *min = g_object_get_data( G_OBJECT(native), "chisel-progress-min" );
	CLFloat *max = g_object_get_data( G_OBJECT(native), "chisel-progress-max" );
	
	CLFloat range = *max - *min;
	CLFloat fraction = gtk_progress_bar_get_fraction( GTK_PROGRESS_BAR(native) );
	
	return *min + (fraction * range);
}

CLFloat _chisel_native_progressbar_get_thickness( native_handle native ) {
	return 25.0; // FIXME: hax
}

void _chisel_native_progressbar_set_indeterminate( native_handle native, int val ) {
	g_object_set_data( G_OBJECT(native), "chisel-progress-indeterminate", (void*)val );
	if ( !val ) {
		g_object_set_data( G_OBJECT(native), "chisel-progress-animating", (void*)0 );
	}
}

int _chisel_native_progressbar_get_indeterminate( native_handle native ) {
	return (int)g_object_get_data( G_OBJECT(native), "chisel-progress-indeterminate" );
}

void _chisel_native_progressbar_set_animating( native_handle native, int val ) {
	int indeterminate = (int)g_object_get_data( G_OBJECT(native), "chisel-progress-indeterminate" );
	
	if ( indeterminate ) {
		if ( val ) {
			g_timeout_add( ANIMATE_UPDATE_INTERVAL, (GSourceFunc)_chisel_native_progressbar_ind_magic, native );
			g_object_set_data( G_OBJECT(native), "chisel-progress-animating", (void*)1 );
		} else {
			g_object_set_data( G_OBJECT(native), "chisel-progress-animating", (void*)0 );
		}
	}
	
	_chisel_force_gtk_refresh( );
}

int _chisel_native_progressbar_get_animating( native_handle native ) {
	return 0;
}
